File: DAA/C Implementation/Week7/mult.c
#include <stdio.h>
#include <limits.h>

int mcm(int arr[], int n)
{
    int dp[n][n];
    for (int i = 1; i < n; i++)
        dp[i][i] = 0;

    for (int i = n - 1; i > 0; i--)
    {
        for (int j = i + 1; j < n; j++)
        {
            dp[i][j] = INT_MAX;
            for (int k = i; k < j; k++)
            {
                int steps = (arr[i - 1] * arr[k] * arr[j]) + dp[i][k] + dp[k + 1][j];
                if (steps < dp[i][j])
                    dp[i][j] = steps;
            }
        }
    }
    return dp[1][n - 1];
}

int main()
{
    int arr[] = {8, 12, 13, 14, 19};
    int n = sizeof(arr) / sizeof(arr[0]);
    int ans = mcm(arr, n);
    printf("Minimal number of scaler multiplications: %d", ans);
}

File: DAA/Python Implementation/Week7/mult.py
import sys

def mcm(arr):
    n = len(arr)
    dp = [[0 for j in range(n)] for i in range(n)]

    for i in range(n-1, 0, -1):
        for j in range(i+1, n):
            dp[i][j] = sys.maxsize
            for k in range(i, j):
                steps = dp[i][k] + dp[k+1][j] + arr[i-1]*arr[k]*arr[j]
                if steps < dp[i][j]:
                    dp[i][j] = steps
    return dp[1][n-1]


arr = [8, 12, 13, 14, 19]
ans = mcm(arr)
print("Minimal number of scaler multiplications:", ans)


#Note: I know the complexity is not bearable but this IS the shortest way to solve this (in terms of lines of code ofc)
